File: src/ValidParentheses.java

public class ValidParentheses {
    public boolean isValid(String s) {
    	java.util.Stack<Character> st = new java.util.Stack<Character>();
    	int len = s.length();
    	
    	for( int i = 0; i < len; i++ ) {
    		if( st.isEmpty() ) {
    			char c = s.charAt(i);
    			st.push(c);
    		}
    		else {
    			char top = st.peek();
    			char c = s.charAt(i);
    			if( match(top, c) )
    				st.pop();
    			else
    				st.push(c);
    		}
    	}
    	if( st.isEmpty() )
    		return true;
    	else 
    		return false;
    }
    private boolean match(char l, char r) {
    	if( l == '(' && r == ')' )
    		return true;
    	else if( l == '{' && r == '}' )
    		return true;
    	else if ( l == '[' && r == ']' )
    		return true;
    	else 
    		return false;
    }
}

File: in_python/valid_parentheses.py
class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return self.items == []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[len(self.items)-1]

    def size(self):
        return len(self.items)


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        # compare later char with previous ones, stack is a tool with spacial order feature
        stack = Stack()
        for p in s:
            if p in "([{":
                stack.push(p)
            else:
                if stack.is_empty():
                    return False
                else:
                    if not self.is_matched(stack.peek(), p):
                        return False
                    else:
                        stack.pop()

        return stack.is_empty()

    def is_matched(self, left, right):
        leftStr = '({['
        rightStr = ')}]'
        return leftStr.index(left) == rightStr.index(right)

if __name__ == "__main__":
    sol = Solution()
    print sol.isValid('{{([][])}()}')
    print sol.isValid('[{()]')
